core-tests/shared/src/test/scala/zio/ChunkSpec.scala

```diff
@@ -459,10 +459,99 @@ object ChunkSpec extends ZIOBaseSpec {
+            flatMap.first eq ZIO.unit
+
+          val doCollect: UIO[Chunk[Int]] = c.collectZIO(pf)
+          for {
+            result <- doCollect
+          } yield assertTrue(
+            result eq Chunk.empty,
+            doCollect match {
+              case flatMap: ZIO.FlatMap[?, ?, ?, ?] =>
+                isSuspendSucceedEncoding(flatMap) &&
+                (flatMap.successK.asInstanceOf[Unit => ZIO[?, ?, ?]].apply(()) eq Exit.emptyChunk)
+              case _ => false
+            }
+          )
+        }
+      },
       test("collectZIO chunk that fails") {
         Chunk(1, 2).collectZIO { case 2 => ZIO.fail("Ouch") }.either.map(assert(_)(isLeft(equalTo("Ouch"))))
-      } @@ zioTag(errors)
-    ),
+      } @@ zioTag(errors),
+      suite("order preservation")(
+        test("preserves order") {
+          for {
+            _    <- ZIO.unit
+            chunk = Chunk(1, 2, 3)
+            collecting <- chunk.collectZIO {
+                            case 1 => ZIO.succeed(1).delay(200.millis).fork
+                            case 2 => ZIO.succeed(2).delay(100.millis).fork
+                            case 3 => ZIO.succeed(3).fork
+                          }
+            result <- ZIO.foreachPar(collecting)(_.join)
+          } yield assertTrue(
+            result == chunk // Order should be preserved in result
+          )
+        } @@ retries(10),
+        test("preserves order with forked ZIOs") {
+          for {
+            _           <- ZIO.unit
+            chunk        = Chunk.fromIterable(1 to 20)
+            randomDelays = Array.fill(20)(scala.util.Random.nextInt(1000).millis)
+            collecting  <- chunk.collectZIO(i => ZIO.attemptBlocking(i * 10).delay(randomDelays(i - 1)).fork)
+            result      <- ZIO.foreachPar(collecting)(_.join)
+          } yield assertTrue(
+            result == chunk.map(_ * 10) // Order should be preserved in result
+          )
+        } @@ retries(10),
+        test("preserves order with mixed sync and async ZIOs") {
+          for {
+            _           <- ZIO.unit
+            chunk        = Chunk.fromIterable(1 to 20)
+            randomDelays = Array.fill(20)(scala.util.Random.nextInt(1000).millis)
+            collecting <- chunk.collectZIO { i =>
+                            val baseEffect = ZIO.attemptBlocking(i * 10)
+                            if (i % 2 == 0) {
+                              // Even numbers: async with fork
+                              baseEffect.delay(randomDelays(i - 1)).fork
+                            } else {
+                              // Odd numbers: sync
+                              baseEffect.fork
+                            }
+                          }
+            result <- ZIO.foreachPar(collecting)(_.join)
+          } yield assertTrue(
+            result == chunk.map(_ * 10) // Order should be preserved in result
+          )
+        } @@ retries(10)
+      ) @@ withLiveClock
```

Comment:
I wasn't sure my new implementation would be able to keep the order of element in the resulting `Chunk` so here I tried to write test to make my implementation fail (return unordered Chunk) but I didn't manange to write test showing that

I'm still not sure my implementation has not an issue with ordering when the effect as async 🤔